File: src/Crawler/Analysis/DnsAnalyzer.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis;

use Crawler\Analysis\Result\DnsAnalysisResult;
use Crawler\Components\SuperTable;
use Crawler\Components\SuperTableColumn;
use Crawler\Options\Options;
use Crawler\Utils;
use Exception;

class DnsAnalyzer extends BaseAnalyzer implements Analyzer
{
    const SUPER_TABLE_DNS = 'dns';

    public function shouldBeActivated(): bool
    {
        return true;
    }

    public function analyze(): void
    {
        $superTable = new SuperTable(
            self::SUPER_TABLE_DNS,
            'DNS info',
            'No DNS info found.', [
            new SuperTableColumn('info', 'DNS resolving tree', 70, function ($value, $renderInto) {
                // find and colorize all IPv4 addresses in $value
                $value = preg_replace_callback('/([0-9]{1,3}\.){3}[0-9]{1,3}/', function ($matches) {
                    $ip = $matches[0];
                    if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
                        return Utils::getColorText($ip, 'blue', true);
                    } else {
                        return $ip;
                    }
                }, $value);

                // find and colorize all IPv6 addresses in $value
                $value = preg_replace_callback('/([0-9a-f:]+:+)+[0-9a-f]+/i', function ($matches) {
                    $ip = $matches[0];
                    if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
                        return Utils::getColorText($ip, 'blue', true);
                    } else {
                        return $ip;
                    }
                }, $value);

                if ($renderInto === SuperTable::RENDER_INTO_HTML) {
                    $value = nl2br(str_replace(' ', '&nbsp;', $value));
                }

                return $value;
            }, null, true, false, true, false),
        ], false);

        $data = [];
        try {
            $dnsInfo = $this->getDnsInfo();
            foreach (explode("\n", $dnsInfo->getTxtDescription()) as $line) {
                $data[] = ['info' => $line];
            }

            $domain = $dnsInfo->resolvedDomains[0] ?? 'unknown';

            // IPv4
            if ($dnsInfo->ipv4Addresses) {
                $this->status->addOkToSummary('dns-ipv4', "DNS IPv4 OK: domain {$domain} resolved to " . implode(', ', $dnsInfo->ipv4Addresses) . " (DNS server: {$dnsInfo->dnsServerName})");
            } else {
                $this->status->addNoticeToSummary('dns-ipv4', "DNS IPv4: domain {$domain} does not support IPv4 (DNS server: {$dnsInfo->dnsServerName})");
            }

            // IPv6
            if ($dnsInfo->ipv6Addresses) {
                $this->status->addOkToSummary('dns-ipv6', "DNS IPv6 OK: domain {$domain} resolved to " . implode(', ', $dnsInfo->ipv6Addresses) . " (DNS server: {$dnsInfo->dnsServerName})");
            } else {
                $this->status->addNoticeToSummary('dns-ipv6', "DNS IPv6: domain {$domain} does not support IPv6 (DNS server: {$dnsInfo->dnsServerName})");
            }

            if (count($dnsInfo->resolvedDomains) > 1) {
                $this->status->addInfoToSummary('dns-aliases', "DNS Aliases: IP(s) for domain {$domain} were resolved by CNAME chain " . implode(' > ', $dnsInfo->resolvedDomains) . '.');
            }

        } catch (Exception $e) {
            $data[] = ['info' => $e->getMessage()];
            $this->status->addCriticalToSummary('dns', "Problem with DNS analysis: {$e->getMessage()}");
        }

        $superTable->setData($data);
        $this->status->addSuperTableAtEnd($superTable);
        $this->output->addSuperTable($superTable);
    }

    /**
     * @return DnsAnalysisResult
     * @throws Exception
     */
    private function getDnsInfo(): DnsAnalysisResult
    {
        $isCygwin = stripos(PHP_OS, 'CYGWIN') !== false;
        $domain = parse_url($this->crawler->getCoreOptions()->url, PHP_URL_HOST);

        // if is not cygwin environment, try to use nslookup or dig
        if (!$isCygwin) {
            $whichNslookup = trim((string)@shell_exec("which nslookup 2>/dev/null"));
            if ($whichNslookup !== '') {
                $nslookupOutput = @shell_exec("nslookup " . escapeshellarg($domain) . ' 2>/dev/null');
                if ($nslookupOutput) {
                    return self::parseNslookup($nslookupOutput);
                }
            }

            $digOutput = @shell_exec("dig +short " . escapeshellarg($domain) . ' 2>/dev/null');
            if ($digOutput) {
                return $this->parseDigOutput($digOutput, $domain);
            }
        }

        // Fallback na dns_get_record
        $records = @dns_get_record($domain, DNS_A + DNS_AAAA + DNS_CNAME);
        if (!$records) {
            throw new \Exception(__METHOD__ . ': Unable to resolve DNS records for ' . $domain);
        }

        $ipv4Addresses = [];
        $ipv6Addresses = [];
        $resolvedDomains = [$domain];

        foreach ($records as $rec) {
            if (isset($rec['type'])) {
                if ($rec['type'] === 'A' && isset($rec['ip'])) {
                    $ipv4Addresses[] = $rec['ip'];
                } elseif ($rec['type'] === 'AAAA' && isset($rec['ipv6'])) {
                    $ipv6Addresses[] = $rec['ipv6'];
                } elseif ($rec['type'] === 'CNAME' && isset($rec['target'])) {
                    $resolvedDomains[] = $rec['target'];
                }
            }
        }

        return new DnsAnalysisResult(
            'unknown-dns-server',
            '0.0.0.0',
            array_unique($resolvedDomains),
            $ipv4Addresses,
            $ipv6Addresses
        );
    }

    public function getOrder(): int
    {
        return 215;
    }

    /**
     * @param string $nslookupOutput
     * @return DnsAnalysisResult
     * @throws Exception
     */
    public static function parseNslookup(string $nslookupOutput): DnsAnalysisResult
    {
        $dnsServerName = null;
        $dnsServerIpAddress = null;
        $resolvedDomains = [];
        $ipv4Addresses = [];
        $ipv6Addresses = [];

        // Extract DNS Server and Address
        if (preg_match('/Server:\s*(\S+)/i', $nslookupOutput, $dnsServerMatches)) {
            $dnsServerName = $dnsServerMatches[1];
        }

        if (preg_match('/Address:\s*([0-9a-z.:]+)/i', $nslookupOutput, $dnsAddressMatches)) {
            $dnsServerIpAddress = $dnsAddressMatches[1];
        }

        if (!$dnsServerName || !$dnsServerIpAddress) {
            throw new Exception('DNS Server or Address not found in nslookup output.');
        }

        // Extract only the "Non-authoritative answer" part
        if (preg_match('/Non-authoritative answer:(.*)/is', $nslookupOutput, $answerMatches)) {
            $answerSection = $answerMatches[1];
        } else {
            $answerSection = $nslookupOutput; // fallback to the entire output if not found
        }

        // Extract IP Addresses from the answer section
        if (preg_match_all('/Address(?:es)?:\s*((?:\S+\s*)+)/is', $answerSection, $ipMatches)) {
            $targetIPs = array_map('trim', preg_split("/\s+/", trim($ipMatches[1][0])));
            foreach ($targetIPs as $ip) {
                if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
                    $ipv4Addresses[] = $ip;
                } else if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
                    $ipv6Addresses[] = $ip;
                }
            }
        }

        // Extract resolved domains from the answer section

        // Windows specific format
        if (preg_match_all('/Aliases:\s*((?:\S+\s*)+)/', $answerSection, $aliasMatchesWindows)) {
            $aliases = explode("\n", trim($aliasMatchesWindows[1][0]));
            foreach ($aliases as $alias) {
                $alias = trim($alias, "\r\n\t .");
                $resolvedDomains[$alias] = $alias;
            }
        }

        // Linux/macOS specific format
        if (preg_match_all('/(\S+)\s+canonical name\s*=\s*(\S+)/i', $answerSection, $aliasMatchesLinux)) {
            foreach ($aliasMatchesLinux[1] as $index => $from) {
                $to = trim($aliasMatchesLinux[2][$index], "\r\n\t .");
                $resolvedDomains[$from] = $from;
                $resolvedDomains[$to] = $to;
            }
        }

        if (preg_match_all('/\s*Name:\s*(\S+)/i', $answerSection, $nameMatches)) {
            foreach ($nameMatches[1] as $match) {
                $match = trim($match, "\r\n\t .");
                $resolvedDomains[$match] = $match;
            }
        }

        if (!$resolvedDomains && !$ipv4Addresses && !$ipv6Addresses) {
            throw new Exception('No resolved domains or IP addresses found. Probably invalid domain.');
        }

        return new DnsAnalysisResult(
            $dnsServerName,
            $dnsServerIpAddress,
            array_unique(array_values($resolvedDomains)),
            $ipv4Addresses,
            $ipv6Addresses
        );
    }

    /**
     * @param string $digOutput
     * @param string $domain
     * @return DnsAnalysisResult
     */
    private function parseDigOutput(string $digOutput, string $domain): DnsAnalysisResult
    {
        $lines = array_filter(array_map('trim', explode("\n", $digOutput)));
        $ipv4Addresses = [];
        $ipv6Addresses = [];
        $resolvedDomains = [$domain];

        foreach ($lines as $line) {
            if (filter_var($line, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
                $ipv4Addresses[] = $line;
            } elseif (filter_var($line, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
                $ipv6Addresses[] = $line;
            } else {
                // Potential CNAME or alias (also remove trailing dot from CNAME)
                $resolvedDomains[] = rtrim($line, '.');
            }
        }

        // DNS server info is not easily obtainable from dig +short, so we set placeholders.
        return new DnsAnalysisResult(
            'unknown-dns-server',
            '0.0.0.0',
            array_unique($resolvedDomains),
            $ipv4Addresses,
            $ipv6Addresses
        );
    }

    public static function getOptions(): Options
    {
        return new Options();
    }
}

File: src/analysis/dns_analyzer.rs
// SiteOne Crawler - DnsAnalyzer
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

use crate::analysis::analyzer::Analyzer;
use crate::analysis::base_analyzer::BaseAnalyzer;
use crate::analysis::result::dns_analysis_result::DnsAnalysisResult;
use crate::components::super_table::SuperTable;
use crate::components::super_table_column::SuperTableColumn;
use crate::output::output::Output;
use crate::result::status::Status;
use crate::utils;

const SUPER_TABLE_DNS: &str = "dns";

pub struct DnsAnalyzer {
    base: BaseAnalyzer,
}

impl Default for DnsAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}

impl DnsAnalyzer {
    pub fn new() -> Self {
        Self {
            base: BaseAnalyzer::new(),
        }
    }

    /// Resolve DNS for the given domain using hickory-resolver.
    fn get_dns_info(&self, domain: &str) -> Result<DnsAnalysisResult, String> {
        use hickory_resolver::Resolver;
        use hickory_resolver::proto::rr::RecordType;

        let domain_owned = domain.to_string();

        // Use block_in_place to allow blocking the current thread while running async DNS lookups
        tokio::task::block_in_place(|| {
            let rt = tokio::runtime::Handle::current();
            rt.block_on(async {
                let resolver = Resolver::builder_tokio()
                    .map_err(|e| format!("Failed to create DNS resolver: {}", e))?
                    .build();

                let mut resolved_domains = vec![domain_owned.clone()];
                let mut ipv4_addresses = Vec::new();
                let mut ipv6_addresses = Vec::new();

                // Resolve CNAME records
                if let Ok(cname_response) = resolver.lookup(domain_owned.as_str(), RecordType::CNAME).await {
                    for record in cname_response.iter() {
                        let cname_str = record.to_string().trim_end_matches('.').to_string();
                        if !resolved_domains.contains(&cname_str) {
                            resolved_domains.push(cname_str);
                        }
                    }
                }

                // Resolve A records (IPv4)
                if let Ok(ipv4_response) = resolver.lookup(domain_owned.as_str(), RecordType::A).await {
                    for record in ipv4_response.iter() {
                        let ip_str = record.to_string();
                        if !ip_str.is_empty() {
                            ipv4_addresses.push(ip_str);
                        }
                    }
                }

                // Resolve AAAA records (IPv6)
                if let Ok(ipv6_response) = resolver.lookup(domain_owned.as_str(), RecordType::AAAA).await {
                    for record in ipv6_response.iter() {
                        let ip_str = record.to_string();
                        if !ip_str.is_empty() {
                            ipv6_addresses.push(ip_str);
                        }
                    }
                }

                if ipv4_addresses.is_empty() && ipv6_addresses.is_empty() {
                    return Err(format!("Unable to resolve DNS records for {}", domain_owned));
                }

                let dns_server_ip = Self::get_system_dns_server().unwrap_or_else(|| "0.0.0.0".to_string());
                let dns_server_name = dns_server_ip.clone();

                Ok(DnsAnalysisResult::new(
                    dns_server_name,
                    dns_server_ip,
                    resolved_domains,
                    ipv4_addresses,
                    ipv6_addresses,
                ))
            })
        })
    }

    /// Read the first nameserver entry from /etc/resolv.conf to get the system DNS server IP.
    fn get_system_dns_server() -> Option<String> {
        let contents = std::fs::read_to_string("/etc/resolv.conf").ok()?;
        for line in contents.lines() {
            let trimmed = line.trim();
            if trimmed.starts_with("nameserver")
                && let Some(ip) = trimmed.split_whitespace().nth(1)
            {
                return Some(ip.to_string());
            }
        }
        None
    }
}

impl Analyzer for DnsAnalyzer {
    fn analyze(&mut self, status: &Status, output: &mut dyn Output) {
        let columns = vec![SuperTableColumn::new(
            "info".to_string(),
            "DNS resolving tree".to_string(),
            70,
            Some(Box::new(|value: &str, _render_into: &str| {
                let mut result = value.to_string();
                // Colorize IPv4 addresses
                if let Ok(re) = regex::Regex::new(r"(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})") {
                    result = re
                        .replace_all(&result, |caps: &regex::Captures| {
                            let ip = &caps[1];
                            if ip.parse::<std::net::Ipv4Addr>().is_ok() {
                                utils::get_color_text(ip, "blue", true)
                            } else {
                                ip.to_string()
                            }
                        })
                        .to_string();
                }
                // Colorize IPv6 addresses
                if let Ok(re) = regex::Regex::new(r"([0-9a-f:]+:+)+[0-9a-f]+") {
                    result = re
                        .replace_all(&result, |caps: &regex::Captures| {
                            let ip = &caps[0];
                            if ip.parse::<std::net::Ipv6Addr>().is_ok() {
                                utils::get_color_text(ip, "blue", true)
                            } else {
                                ip.to_string()
                            }
                        })
                        .to_string();
                }
                result
            })),
            None,
            true,
            false,
            true,
            false,
            None,
        )];

        let mut super_table = SuperTable::new(
            SUPER_TABLE_DNS.to_string(),
            "DNS info".to_string(),
            "No DNS info found.".to_string(),
            columns,
            false,
            None,
            "ASC".to_string(),
            None,
            None,
            None,
        );

        let mut data: Vec<HashMap<String, String>> = Vec::new();

        // Extract domain from the first visited URL
        let domain = status
            .get_visited_urls()
            .first()
            .and_then(|u| u.get_host())
            .unwrap_or_else(|| "unknown".to_string());

        match self.get_dns_info(&domain) {
            Ok(dns_info) => {
                for line in dns_info.get_txt_description().lines() {
                    let mut row = HashMap::new();
                    row.insert("info".to_string(), line.to_string());
                    data.push(row);
                }

                let resolved_domain = dns_info
                    .resolved_domains
                    .first()
                    .cloned()
                    .unwrap_or_else(|| "unknown".to_string());

                // DNS server suffix — omit when unknown (e.g. on Windows where /etc/resolv.conf doesn't exist)
                let dns_suffix = if dns_info.dns_server_ip_address != "0.0.0.0" {
                    format!(" (DNS server: {})", dns_info.dns_server_name)
                } else {
                    String::new()
                };

                // IPv4 summary
                if !dns_info.ipv4_addresses.is_empty() {
                    status.add_ok_to_summary(
                        "dns-ipv4",
                        &format!(
                            "DNS IPv4 OK: domain {} resolved to {}{}",
                            resolved_domain,
                            dns_info.ipv4_addresses.join(", "),
                            dns_suffix
                        ),
                    );
                } else {
                    status.add_notice_to_summary(
                        "dns-ipv4",
                        &format!(
                            "DNS IPv4: domain {} does not support IPv4{}",
                            resolved_domain, dns_suffix
                        ),
                    );
                }

                // IPv6 summary
                if !dns_info.ipv6_addresses.is_empty() {
                    status.add_ok_to_summary(
                        "dns-ipv6",
                        &format!(
                            "DNS IPv6 OK: domain {} resolved to {}{}",
                            resolved_domain,
                            dns_info.ipv6_addresses.join(", "),
                            dns_suffix
                        ),
                    );
                } else {
                    status.add_notice_to_summary(
                        "dns-ipv6",
                        &format!(
                            "DNS IPv6: domain {} does not support IPv6{}",
                            resolved_domain, dns_suffix
                        ),
                    );
                }

                // CNAME chain summary
                if dns_info.resolved_domains.len() > 1 {
                    status.add_info_to_summary(
                        "dns-aliases",
                        &format!(
                            "DNS Aliases: IP(s) for domain {} were resolved by CNAME chain {}.",
                            resolved_domain,
                            dns_info.resolved_domains.join(" > ")
                        ),
                    );
                }
            }
            Err(e) => {
                let mut row = HashMap::new();
                row.insert("info".to_string(), e.clone());
                data.push(row);
                status.add_critical_to_summary("dns", &format!("Problem with DNS analysis: {}", e));
            }
        }

        super_table.set_data(data);
        status.configure_super_table_url_stripping(&mut super_table);
        output.add_super_table(&super_table);
        status.add_super_table_at_end(super_table);
    }

    fn should_be_activated(&self) -> bool {
        true
    }

    fn get_order(&self) -> i32 {
        215
    }

    fn get_name(&self) -> &str {
        "DnsAnalyzer"
    }

    fn get_exec_times(&self) -> &HashMap<String, f64> {
        self.base.get_exec_times()
    }

    fn get_exec_counts(&self) -> &HashMap<String, usize> {
        self.base.get_exec_counts()
    }
}
